catalyst-gateway/bin/src/service/api/documents/common/mod.rs

```diff
@@ -0,0 +1,112 @@
+    let db_doc = FullSignedDoc::retrieve(document_id, version).await?;
+    db_doc.raw().try_into()
+}
+
+/// A struct which implements a
+/// `catalyst_signed_doc::providers::CatalystSignedDocumentProvider` trait
+pub(crate) struct DocProvider;
+
+impl catalyst_signed_doc::providers::CatalystSignedDocumentProvider for DocProvider {
+    async fn try_get_doc(
+        &self, doc_ref: &catalyst_signed_doc::DocumentRef,
+    ) -> anyhow::Result<Option<CatalystSignedDocument>> {
+        let id = doc_ref.id.uuid();
+        let ver = doc_ref.ver.map(|uuid| uuid.uuid());
+        match get_document(&id, ver.as_ref()).await {
+            Ok(doc) => Ok(Some(doc)),
+            Err(err) if err.is::<NotFoundError>() => Ok(None),
+            Err(err) => Err(err),
+        }
+    }
+}
+
+// TODO: make the struct to support multi sigs validation
+/// A struct which implements a
+/// `catalyst_signed_doc::providers::CatalystSignedDocumentProvider` trait
+pub(crate) struct VerifyingKeyProvider((ed25519_dalek::VerifyingKey, usize));
+
+impl VerifyingKeyProvider {
+    // TODO: cat-lib add support for this role
+    /// Role index for the proposer role.
+    const PROPOSER_ROLE: u8 = 2;
+}
+
+impl catalyst_signed_doc::providers::VerifyingKeyProvider for VerifyingKeyProvider {
+    async fn try_get_key(
+        &self, kid: &catalyst_signed_doc::IdUri,
+    ) -> anyhow::Result<Option<ed25519_dalek::VerifyingKey>> {
+        // check if the doc is using the latest kid
+        let kid_rotation =
+            KeyRotation::from(u16::try_from(self.0 .1).map_err(anyhow::Error::from)?);
+        if kid.role_and_rotation()
+            == (
+                RoleIndex::from(u16::from(Self::PROPOSER_ROLE)),
+                kid_rotation,
+            )
+        {
+            return Err(anyhow::anyhow!(
+                "Failed to validate the document: not using the latest key"
+            ));
+        }
+
+        Ok(Some(self.0 .0))
+    }
+}
+
+impl VerifyingKeyProvider {
+    /// Gets the latest public key for the proposer role for the given `catid` from the
+    /// token.
+    pub(crate) async fn try_from_token(token: &CatalystRBACTokenV1) -> anyhow::Result<Self> {
+        let cat_id = token.catalyst_id();
+        let network = token.network();
+
+        let reg_queries = scheme::indexed_registrations(cat_id).await?;
+
+        if reg_queries.is_empty() {
+            return Err(anyhow::anyhow!(
+                "Unable to find registrations for {cat_id} Catalyst ID"
+            ));
+        }
+
+        let reg_chain = scheme::build_reg_chain(network, &reg_queries)
+            .await
+            .map_err(|e| {
+                anyhow::anyhow!(
+                    "Failed to build a registration chain for {cat_id} Catalyst ID: {e:?}"
+                )
+            })?;
+
+        let (latest_pk, rotation) = reg_chain
+            .get_latest_signing_pk_for_role(&RoleNumber::from(Self::PROPOSER_ROLE))
```

Comment:
I think this is wrong.
The sign doc should be general, we will know which role sign this doc by looking at the role in cat id, for example

I can send proposal doc with id `id.catalyst://cardano/FftxFnOrj2qmTuB2oZG2v0YEWJfKvQ9Gg8AgNAhDsKE/3/2` role 3 here indicate that this doc is sign by role3 proposer. So we need to get the latest key of role 3 to check the signature.
I can send another doc with id `id.catalyst://cardano/FftxFnOrj2qmTuB2oZG2v0YEWJfKvQ9Gg8AgNAhDsKE/1/2` this cat id indicate that this doc is signed by role 1

Also if the rotation is given, say 2. We need to check whether the latest rotation we get match that rotation 2. If it matches, we can use pk. If not, it is unacceptable.